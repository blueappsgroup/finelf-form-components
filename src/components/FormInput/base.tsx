/* eslint-disable @typescript-eslint/no-explicit-any */
import React, { ReactElement, useContext, ChangeEvent } from 'react'
import styled from 'styled-components'
import { FieldWrapProps, SelectFieldWrapProps } from '../../types'
import { device } from '../../consts/sizes'
import { FormContext, setFormValuesToCache } from '../../utils'

type StyledProps = {
  theme: {
    inputErrorColor: string
    inputBgColor: string
    inputPlaceHolderColor: string
    inputBorderColor: string
    inputTextColor: string
  }
  error?: string | boolean
}

const StyledRow = styled.div`
  display: flex;
  align-items: center;
  margin-bottom: 12px;
  position: relative;
`

export const Row = styled.div`
  display: flex;
  flex-direction: column;

  @media ${device.tablet} {
    flex-direction: row;
    & ${StyledRow}:first-of-type {
      flex-grow: 1;
      margin-right: 5px;
    }

    & ${StyledRow}:last-child {
      flex-grow: 1;
      margin-left: 5px;
    }
  }
`

const StyledError = styled.span`
  color: ${(props: StyledProps): string => props.theme.inputErrorColor};
  position: absolute;
  font-size: 10px;
  bottom: -12px;
`

const StyledInput = styled.input<any>`
  background: ${(props: StyledProps): string => props.theme.inputBgColor};
  border: 1px solid
    ${(props: StyledProps): string => props.theme.inputBorderColor};
  box-sizing: border-box;
  box-shadow: 0px 7px 64px rgba(194, 186, 186, 0.07);
  border-radius: 6px;
  width: 100%;
  height: 63px;
  display: flex;
  align-items: center;
  font-style: normal;
  font-weight: 500;
  font-size: 14px;
  line-height: 24px;
  padding: 0px 20px;
  border-color: ${(props: StyledProps): string =>
    props.error ? props.theme.inputErrorColor : props.theme.inputBorderColor};
  color: ${(props: any): string =>
    props.error ? props.theme.inputErrorColor : props.theme.inputTextColor};

  &::placeholder {
    color: ${(props: StyledProps): string => props.theme.inputPlaceHolderColor};
  }

  &:focus {
    outline: none;
  }
`

const StyledSelect = styled.select<any>`
  background: ${(props: StyledProps): string => props.theme.inputBgColor};
  border: 1px solid
    ${(props: StyledProps): string => props.theme.inputBorderColor};
  box-sizing: border-box;
  box-shadow: 0px 7px 64px rgba(194, 186, 186, 0.07);
  border-radius: 6px;
  width: 100%;
  height: 63px;
  display: flex;
  align-items: center;
  font-style: normal;
  font-weight: 500;
  font-size: 14px;
  line-height: 24px;
  padding: 0px 20px;
  border-color: ${(props: StyledProps): string =>
    props.error ? props.theme.inputErrorColor : props.theme.inputBorderColor};
  color: ${(props: any): string =>
    props.error ? props.theme.inputErrorColor : props.theme.inputTextColor};

  &::placeholder {
    color: ${(props: StyledProps): string => props.theme.inputPlaceHolderColor};
  }

  &:focus {
    outline: none;
  }
`

const BaseField: (props: FieldWrapProps) => ReactElement = ({
  field,
  form: { touched, errors, values },
  ...props
}) => {
  const { id } = useContext(FormContext)
  const handleOnBlur = (e: ChangeEvent<HTMLInputElement>): void => {
    setFormValuesToCache(values, id)
    field.onBlur && field.onBlur(e)
  }

  return (
    <StyledRow>
      {props.label && <label htmlFor={field.name}>{props.label}</label>}
      <StyledInput
        {...field}
        {...props}
        onBlur={handleOnBlur}
        type={props.type}
        value={(field.value && field.value) || ''}
        error={touched[field.name] && errors[field.name]}
        placeholder={
          props.placeholder &&
          `${props.placeholder}${(props.required && '*') || ''}`
        }
      />
      {props.showError && touched[field.name] && errors[field.name] && (
        <StyledError>{errors[field.name]}</StyledError>
      )}
    </StyledRow>
  )
}

export default BaseField

export const BaseSelectField: (props: SelectFieldWrapProps) => ReactElement = ({
  field,
  form: { touched, errors, values },
  ...props
}) => {
  const { id } = useContext(FormContext)
  const handleOnBlur = (e: ChangeEvent<HTMLInputElement>): void => {
    setFormValuesToCache(values, id)
    field.onBlur && field.onBlur(e)
  }

  const options = (options: []): Array<JSX.Element> => {
    const opts = []

    for (const option in options) {
      opts.push(
        <option key={option} value={option}>
          {options[option]}
        </option>
      )
    }

    return opts
  }

  return (
    <StyledRow>
      {props.label && <label htmlFor={field.name}>{props.label}</label>}
      <StyledSelect
        {...field}
        {...props}
        onBlur={handleOnBlur}
        type="text"
        value={(field.value && field.value) || ''}
        error={touched[field.name] && errors[field.name]}
        placeholder={
          props.placeholder &&
          `${props.placeholder}${(props.required && '*') || ''}`
        }
      >
        {props.options && options(props.options)}
      </StyledSelect>
      {props.showError && touched[field.name] && errors[field.name] && (
        <StyledError>{errors[field.name]}</StyledError>
      )}
    </StyledRow>
  )
}
