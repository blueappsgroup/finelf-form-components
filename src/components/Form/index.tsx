import React, { FC, ReactElement, useMemo, useState } from 'react'
import { Formik, Form } from 'formik'
import styled from 'styled-components'

import { ThemeProvider } from '../../consts/theme'
import { device } from '../../consts/sizes'
import { FormProps, FormValuesType } from '../../types'
import RedirectPage from '../RedirectPage'
import {
  FormContext,
  getFormValuesFromCache,
  resetFormValueCache,
  handleSendDataToApi,
  getFieldsValuesFromUrl,
} from '../../utils'
import { formStatuses } from '../../consts/form'

const StyledForm = styled(Form)`
  display: flex;
  flex-direction: column;
  max-width: ${(props): string => props.theme.formMaxWidth};
  justify-self: center;
  background: ${(props): string => props.theme.formBgColor};
  font-family: ${(props): string => props.theme.fontFamily};
  border-radius: 6px;
  box-shadow: ${(props): string => props.theme.formBoxShadow};
`

const FormWrapper: FC<FormProps> = ({
  children,
  onSubmit,
  customTheme,
  id,
  stepsLength,
  stepsTitles,
  hasRedirect,
  redirectUrl,
  timeToRedirect,
  redirectHeaderText,
  logoImg,
  redirectMainImg,
  redirectBgImg,
  sendDataToApi,
  apiUrl,
  transactionName,
  propertyNamesFromUrl,
  dataWithUserAgent,
}) => {
  const trasationIdValue =
    transactionName &&
    new URLSearchParams(window.location.search).get(transactionName)
  const intialValuesFromUrl = useMemo(
    () =>
      (propertyNamesFromUrl &&
        propertyNamesFromUrl.length > 0 &&
        getFieldsValuesFromUrl(propertyNamesFromUrl)) ||
      {},
    [propertyNamesFromUrl]
  )
  const [redirectUrlPath, setRedirectUrlPath] = useState(redirectUrl)
  const [initialValues, setInitialValues] = useState({
    ...intialValuesFromUrl,
    ...getFormValuesFromCache(id),
    // eslint-disable-next-line @typescript-eslint/camelcase
    trasaction_id: trasationIdValue,
  })
  const [currentStep, setCurrentStep] = useState(0)
  const [fieldsForSkip, setFieldsForSkip] = useState<string[]>([])
  const [errorFromApi, setErrorFromApi] = useState<boolean>(false)
  const shouldRedirect = !errorFromApi && hasRedirect
  const addFieldForSkip = (key: string): void =>
    setFieldsForSkip([...fieldsForSkip, key])

  const handleSubmit = async (
    values: FormValuesType,
    props: {
      resetForm: () => void
      setStatus: (status: string) => void
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ): Promise<any> => {
    if (onSubmit) {
      onSubmit(values, props)
    }

    try {
      if (sendDataToApi && apiUrl) {
        const response = await handleSendDataToApi(
          values,
          apiUrl,
          id,
          fieldsForSkip,
          dataWithUserAgent
        )
        const {
          redirectUrl: urlFromApi,
          status: statusFromApi,
        } = await response.json()

        if (statusFromApi === false) {
          setErrorFromApi(true)
          return
        }

        urlFromApi && setRedirectUrlPath(urlFromApi)
      }

      props.resetForm()
      props.setStatus(formStatuses.submited)
    } catch (e) {
      props.setStatus(formStatuses.error)
    }
  }
  const handleReset = (): void => {
    resetFormValueCache(id)
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    setInitialValues({} as any)
  }

  const prevStep: Function = () => setCurrentStep(currentStep - 1)

  const nextStep: Function = () => setCurrentStep(currentStep + 1)

  return (
    <FormContext.Provider
      value={{
        id,
        apiUrl,
        stepsLength,
        currentStep,
        stepsTitleList: stepsTitles,
        nextStep,
        prevStep,
        fieldsForSkip,
        addFieldForSkip,
        errorFromApi,
      }}
    >
      <ThemeProvider customTheme={{ ...customTheme }}>
        <Formik
          enableReinitialize
          initialValues={initialValues}
          onSubmit={handleSubmit}
          onReset={handleReset}
        >
          {(props): ReactElement =>
            (shouldRedirect && props.status === formStatuses.submited && (
              <RedirectPage
                redirectUrl={redirectUrlPath}
                backgroundImage={redirectBgImg}
                logoImg={logoImg}
                headerText={redirectHeaderText}
                timeToRedirect={timeToRedirect}
                mainImg={redirectMainImg}
              />
            )) || (
              <StyledForm id={id}>
                {children}
              </StyledForm>
            )
          }
        </Formik>
      </ThemeProvider>
    </FormContext.Provider>
  )
}

export default FormWrapper
