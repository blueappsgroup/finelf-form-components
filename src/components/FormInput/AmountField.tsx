import React, { FC } from 'react'
import { Field } from 'formik'

import { RangeFieldProps } from '../../types'
import { BaseRangeField } from './base'

const AmountField: FC<RangeFieldProps> = ({
  id,
  label = 'Kwota pożyczki',
  name,
  required = false,
  showError = true,
  showMinMax = true,
  min = 300,
  max = 10000,
  step = 100,
  value = 2000,
  suffix = 'PLN',
  unit = 'zł',
}) => (
  <Field
    required={required}
    showError={showError}
    showMinMax={showMinMax}
    id={id}
    name={name}
    component={BaseRangeField}
    label={label}
    type="text"
    min={min}
    max={max}
    step={step}
    value={value}
    suffix={suffix}
    unit={unit}
  />
)

AmountField.defaultProps = {
  name: 'amount',
}

export default AmountField
