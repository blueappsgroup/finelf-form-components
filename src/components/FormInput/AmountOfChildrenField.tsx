import React, { ReactElement } from 'react'
import { Field } from 'formik'

import BaseField from './base'
import { validateAmountOfChildren } from './validate'
import { AmountOfChildrenFieldProps } from '../../types'

const AmountOfChildrenField: (
  props: AmountOfChildrenFieldProps
) => ReactElement = ({
  id,
  label,
  placeholder = 'AmountOfChildren',
  errorMsg = 'Podana ilość jest nieprawidłowa',
  name = 'amount_of_children',
  required = false,
  showError,
}) => (
  <Field
    required={required}
    showError={showError}
    id={id}
    name={name}
    component={BaseField}
    label={label}
    placeholder={placeholder}
    type="text"
    validate={validateAmountOfChildren(errorMsg, required)}
  />
)

export default AmountOfChildrenField
