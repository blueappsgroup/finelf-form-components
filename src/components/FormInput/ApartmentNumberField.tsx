import React from 'react'
import TextField from './TextField'
import { validateHouseNumber } from './validate'

const ApartmentNumberField = () => (
  <TextField
    required
    label="Numer domu"
    placeholder="Numer domu"
    showError
    name="house_number"
    validate={validateHouseNumber}
  />
)

export default ApartmentNumberField
