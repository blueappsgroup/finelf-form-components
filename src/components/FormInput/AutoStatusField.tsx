import { Field } from 'formik'
import React, { ReactElement } from 'react'
import { SelectFieldProps } from '../../types'

import { BaseSelectField } from './base'

const options = {
  yes: 'tak',
  no: 'nie',
}

const AutoStatusField: (props: SelectFieldProps) => ReactElement = ({
  id,
  label,
  name = 'auto_status',
  required = false,
  showError,
}) => (
  <Field
    required={required}
    showError={showError}
    id={id}
    name={name}
    component={BaseSelectField}
    label={label}
    type="select"
    options={options}
  />
)

export default AutoStatusField
