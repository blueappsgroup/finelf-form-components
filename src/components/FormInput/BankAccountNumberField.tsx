import React, { FC } from 'react'
import { Field } from 'formik'

import { TextFieldProps } from '../../types'
import BaseField from './base'
import { validateBankAccountNumber } from './validate'

const BankAccountNumberField: FC<TextFieldProps> = ({
  id,
  label = 'Numer konta bankowego',
  placeholder,
  name,
  required = false,
  validate = validateBankAccountNumber,
  showError = true,
}) => (
  <Field
    id={id}
    label={label}
    placeholder={placeholder}
    component={BaseField}
    required={required}
    showError={showError}
    name={name}
    validate={validate(required)}
  />
)

BankAccountNumberField.defaultProps = {
  name: 'bank_account_number',
}

export default BankAccountNumberField
