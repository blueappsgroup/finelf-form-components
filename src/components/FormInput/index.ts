export { default as EmailField } from './EmailField'
export { default as TextField } from './TextField'
export { default as TelephoneField } from './TelephoneField'
export { default as CheckboxField } from './CheckboxField'
export { default as FirstNameField } from './FirstNameField'
export { default as LastNameField } from './LastNameField'
export { default as PeselField } from './PeselField'
export { default as HouseNumberField } from './HouseNumberField'
export { default as ApartmentNumberField } from './ApartmentNumberField'
export { default as CityNameField } from './CityNameField'
export { default as StreetNameField } from './StreetNameField'
export { default as ZipCodeField } from './ZipCodeField'
export { default as CompanyNameField } from './CompanyNameField'
export { default as CompanyTaxField } from './CompanyTaxField'
export { default as BankAccountNumberField } from './BankAccountNumberField'
export { default as IDCardNumberField } from './IDCardNumberField'
export { Row } from './base'
