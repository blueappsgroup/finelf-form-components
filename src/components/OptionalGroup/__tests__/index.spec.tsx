import * as React from 'react'
import { render, fireEvent, act } from '@testing-library/react'

import Form from '../../Form'
import { FirstNameField, LastNameField } from '../../FormInput'
import { OptionalGroup } from '../../index'

describe('base <OptionalGroup />', () => {
  const onSubmit = jest.fn()
  const setupWrapper = ({
    formId = 'testForm',
    label = 'optional group test',
    name = 'optional_group',
    ...rest
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  }): any => {
    const wrapper = render(
      <Form id={formId} onSubmit={onSubmit}>
        <OptionalGroup name={name} label={label}>
          <FirstNameField name="fName" />
          <LastNameField name="lName" />
        </OptionalGroup>
      </Form>
    )
    const { container } = wrapper

    const optionalGroup = container.querySelector(`[name="${name}"]`)

    return {
      optionalGroup,
      ...wrapper,
    }
  }

  it('matches snapshot', () => {
    const { optionalGroup } = setupWrapper({})

    expect(optionalGroup).toBeTruthy()
    expect(optionalGroup).toMatchSnapshot()
  })
})
