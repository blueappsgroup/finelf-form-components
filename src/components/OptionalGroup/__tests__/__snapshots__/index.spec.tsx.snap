// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`base <OptionalGroup /> matches snapshot 1`] = `
<input
  class="sc-kLokih QhdbU"
  id="optional_group"
  label="optional group test"
  name="optional_group"
  type="checkbox"
  value="false"
/>
`;
